### ISSUE TITLE
[package] cpprestsdk/2.10.16: private boost require needed

### ISSUE THREAD

**Issue by wpalfi** [conan-io/conan-center-index#3285]
I need to use cpprestsdk together with a boost installation coming not from conan. Would be no problem if cpprestsdk would use a private require, as it uses boost only internally. 

I saw that there is already a component `cpprestsdk_boost_internal`. Is it meant to hide the boost require?

Boost versions are not binary compatible, so version overriding is in general dangerous.

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **cpprestsdk/2.10.16**
  * Operating System+version: **Windows**
  * Compiler+version: **Visual Studio 15 2017**
  * Conan version: **conan 1.30.1**
  * Python version: **Python 3.6**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=Visual Studio
compiler.version=15
build_type=Release
[options]
[build_requires]
[env]
```

**Comment by madebr** [comment]
When building cpprestsdk as a static library, you will also need to link against the boost/zlib/websocketpp libraries.
Granted: when building a shared cpprestsdk library, it is not necessary.

So the internal boost requirement should only be added to the *final* cpprestsdk component when building a static library.
I don't think there exists a method with conan to only propagate a requirement when building a static library (or another condition).

I can be done, with some `if not self.options.shared`.
```python
def package_info(self):
    # cpprestsdk_boost_internal
    self.cpp_info.components["cpprestsdk_boost_internal"].includedirs = []
    self.cpp_info.components["cpprestsdk_boost_internal"].requires = ["boost::boost"]
    # cpprestsdk_openssl_internal
    self.cpp_info.components["cpprestsdk_openssl_internal"].includedirs = []
    self.cpp_info.components["cpprestsdk_openssl_internal"].requires = ["openssl::openssl"]
    # cpprest
    self.cpp_info.components["cpprest"].libs = tools.collect_libs(self)
    if not self.options.shared:
        self.cpp_info.components["cpprest"].requires = ["cpprestsdk_boost_internal",
                                                        "cpprestsdk_openssl_internal"]
    if self.settings.os == "Linux":
        self.cpp_info.components["cpprest"].system_libs.append("pthread")
    elif self.settings.os == "Windows":
        if self.options.get_safe("http_client_impl") == "winhttp":
            self.cpp_info.components["cpprest"].system_libs.append("winhttp")
        if self.options.get_safe("http_listener_impl") == "httpsys":
            self.cpp_info.components["cpprest"].system_libs.append("httpapi")but it would result in code bloat:

        self.cpp_info.components["cpprest"].system_libs.append("bcrypt")
        if self.options.get_safe("pplx_impl") == "winpplx":
            self.cpp_info.components["cpprest"].defines.append("CPPREST_FORCE_PPLX=1")
        if self.options.get_safe("http_client_impl") == "asio":
            self.cpp_info.components["cpprest"].defines.append("CPPREST_FORCE_HTTP_CLIENT_ASIO")
        if self.options.get_safe("http_listener_impl") == "asio":
            self.cpp_info.components["cpprest"].defines.append("CPPREST_FORCE_HTTP_LISTENER_ASIO")
    elif self.settings.os == "Macos":
        self.cpp_info.components["cpprest"].frameworks.extend(["CoreFoundation", "Security"])
    if not self.options.shared:
        self.cpp_info.components["cpprest"].defines.append("_NO_ASYNCRTIMP")
    # cpprestsdk_zlib_internal
    if self.options.with_compression:
        self.cpp_info.components["cpprestsdk_zlib_internal"].includedirs = []
        self.cpp_info.components["cpprestsdk_zlib_internal"].requires = ["zlib::zlib"]
        if not self.options.shared:
            self.cpp_info.components["cpprest"].requires.append("cpprestsdk_zlib_internal")
    # cpprestsdk_websocketpp_internal
    if self.options.with_websockets:
        self.cpp_info.components["cpprestsdk_websocketpp_internal"].includedirs = []
        self.cpp_info.components["cpprestsdk_websocketpp_internal"].requires = ["websocketpp::websocketpp"]
        if not self.options.shared:
            self.cpp_info.components["cpprest"].requires.append("cpprestsdk_websocketpp_internal")

```

The question is whether this added complexity is desired?
Maybe conan does have functionality for this, or is it in the pipeline?

@uilianries @danimtb @SpaceIm @prince-chrismc 
What is your view on this?

**Comment by SpaceIm** [comment]
I think that conan should provide properties for components as private/public/interface dependency, shared/static etc, and handle propagation of dependencies libs/flags internally based on these informations.

**Comment by prince-chrismc** [comment]
I agree with what you guys have said, I am just not convinced the bug is a Conan issue 🤔 you want to install cpprestsdk via Conan to build with or ship/deploy your application?

If you are building, there's no way to exclude an include, but you can choose to delete the file from the generator for instance with cmake_find_package.

If you shipping/deploying your pipeline could simply exclude boost when copying runtimes knowing it installed the correct version for another source.

Can you please explain what you are trying to accomplish so we can provide better feedback on your issue to help resolve it?

**Comment by danimtb** [comment]
I think we previously bumped into this issue and we created a package ID mode for that https://docs.conan.io/en/latest/reference/conanfile/methods.html#self-info-shared-library-package-id. However, we were not sure of introducing that in ConanCenter as it is not the Conan default and it has not been widely tested.

As you comment, there are some things missing in the current dependency model of Conan that we are aware, but we are focusing on Conan 2.0 to improve the graph model and its computation.

So, although the current model it is not accurate enough, let's make the current recipes in CCI work together nicely as we currently cannot cover the use case described in the issue.

**Comment by SpaceIm** [comment]
@danimtb Is it not another issue? `self.info.shared_library_package_id()` doesn't address public/private property of dependencies.

There was a question long time ago about `self.info.shared_library_package_id()` https://github.com/conan-io/conan-center-index/issues/1609. I would love to have this feature in CCI.

**Comment by danimtb** [comment]
@SpaceIm yes, it is not the same cases but I was trying to say that the issues exposed are similar in the sense that the current graph model should be improved in order to properly support the cases described here as well as the cases that `shared_library_package_id()` tried to mitigate.
